### ISSUE TITLE
Terminal Size Zero in Docker Run

### ISSUE THREAD

**Issue by MoritzKn** [moby/moby#43229]
**Description**

`ioctl(STDIN_FILENO, TIOCGWINSZ, &ws)` returns a terminal size of 0 columns and 0 lines and no error code when run in docker.

When using `tput` you see 24 lines and 80 columns. I think the underlying issue is the same but `tput` seems to fall back to a default value.

This is only the case when the command is run right away after starting the container with `docker run`. A short delay fixes the issue.

**Steps to reproduce the issue:**

See this repo for a detailed minimal reproduction:
- [MoritzKn/docker-terminal-size-issues](https://github.com/MoritzKn/docker-terminal-size-issues)

Or this code snipped for a quick demo of the issue:

```sh
docker run -t ubuntu:16.04 tput lines
# vs
docker run -t ubuntu:16.04 bash -c 'sleep 0.1 && tput lines'
```

**Describe the results you received:**
The first command always returns 24. The second command returns the actual terminal size.

**Describe the results you expected:**
Both commands should return the same value, i.e. the actual terminal size.

**Additional information you deem important (e.g. issue happens only occasionally):**
This was already reported previously as #25450 and supposed to be fixed through opencontainers/runc#1275 and #37172 however the behavior seems to have persisted or reappeared.

This is also evidenced by [this comment](https://github.com/moby/moby/issues/25450#issuecomment-357355857) with 14 likes:
> I just updated to the latest CE docker for mac install and I ran into this issue. I'm only able to get it to "work" by passing in the COLUMNS/LINES/TERM vars into docker exec.
 >Is this something that hasn't landed yet on docker for mac?

**Output of `docker version`:**

```
Client:
 Cloud integration: 1.0.17
 Version:           20.10.8
 API version:       1.41
 Go version:        go1.16.6
 Git commit:        3967b7d
 Built:             Fri Jul 30 19:55:20 2021
 OS/Arch:           darwin/arm64
 Context:           default
 Experimental:      true

Server: Docker Engine - Community
 Engine:
  Version:          20.10.8
  API version:      1.41 (minimum version 1.12)
  Go version:       go1.16.6
  Git commit:       75249d8
  Built:            Fri Jul 30 19:53:48 2021
  OS/Arch:          linux/arm64
  Experimental:     true
 containerd:
  Version:          1.4.9
  GitCommit:        e25210fe30a0a703442421b0f60afac609f950a3
 runc:
  Version:          1.0.1
  GitCommit:        v1.0.1-0-g4144b63
 docker-init:
  Version:          0.19.0
  GitCommit:        de40ad0
```

**Output of `docker info`:**

```
Client:
 Context:    default
 Debug Mode: false
 Plugins:
  buildx: Build with BuildKit (Docker Inc., v0.6.3)
  compose: Docker Compose (Docker Inc., v2.0.0)
  scan: Docker Scan (Docker Inc., v0.8.0)

Server:
 Containers: 313
  Running: 0
  Paused: 0
  Stopped: 313
 Images: 126
 Server Version: 20.10.8
 Storage Driver: overlay2
  Backing Filesystem: extfs
  Supports d_type: true
  Native Overlay Diff: true
  userxattr: false
 Logging Driver: json-file
 Cgroup Driver: cgroupfs
 Cgroup Version: 1
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
 Swarm: inactive
 Runtimes: io.containerd.runc.v2 io.containerd.runtime.v1.linux runc
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: e25210fe30a0a703442421b0f60afac609f950a3
 runc version: v1.0.1-0-g4144b63
 init version: de40ad0
 Security Options:
  seccomp
   Profile: default
 Kernel Version: 5.10.47-linuxkit
 Operating System: Docker Desktop
 OSType: linux
 Architecture: aarch64
 CPUs: 4
 Total Memory: 3.841GiB
 Name: docker-desktop
 ID: C73Z:I6SH:3JDZ:UE3T:MK7K:27OA:V4RI:ZIF7:PEAA:F33M:BDMB:IION
 Docker Root Dir: /var/lib/docker
 Debug Mode: false
 HTTP Proxy: http.docker.internal:3128
 HTTPS Proxy: http.docker.internal:3128
 Registry: https://index.docker.io/v1/
 Labels:
 Experimental: true
 Insecure Registries:
  127.0.0.0/8
 Live Restore Enabled: false
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**

Reproduced both on Docker for Mac and Docker for Linux.

**Comment by corhere** [comment]
This should be fixed by #43593. Could you please try reproducing the issue on v22.06.0-beta.0?

**Comment by thaJeztah** [comment]
Yup! Looks resolved on 22.06.0-beta.0;

```bash
docker run -t --rm ubuntu:16.04 tput lines
37

docker run -t --rm ubuntu:16.04 tput cols
204
```

**Comment by AlexPykavy** [comment]
@thaJeztah , I've faced exactly the same issue. It lead to incorrect `mono` work which uses similar to @MoritzKn C logic to set terminal sizes.

Moreover, I've found that `docker run -t --rm ubuntu:20.04 tput lines` works on both my current Docker Engine and v22.06.0-beta.0. However, the C app returns non-zero size only on the beta Docker Engine.

My current Docker Engine version is:
```
Client:
 Version:           20.10.18
 API version:       1.41
 Go version:        go1.18.6
 Git commit:        b40c2f6
 Built:             Thu Sep  8 23:17:48 2022
 OS/Arch:           windows/amd64
 Context:           default
 Experimental:      true

Server: Docker Desktop 4.12.0 (85629)
 Engine:
  Version:          20.10.17
  API version:      1.41 (minimum version 1.12)
  Go version:       go1.17.11
  Git commit:       a89b842
  Built:            Mon Jun  6 23:01:23 2022
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          1.6.8
  GitCommit:        9cd3357b7fd7218e4aec3eae239db1f68a5a6ec6
 runc:
  Version:          1.1.4
  GitCommit:        v1.1.4-0-g5fd4c4d
 docker-init:
  Version:          0.19.0
  GitCommit:        de40ad0
```

Do you know when 22 Docker Engine is going to go live? And is it possible to backport this fix to 20.X?
